lib/ruby_indexer/lib/ruby_indexer/declaration_listener.rb

```diff
@@ -67,6 +67,8 @@ def on_class_node_enter(node)
       parent_class = case superclass
       when Prism::ConstantReadNode, Prism::ConstantPathNode
         superclass.slice
+      else
+        "Object"
```

Comment:
I'm not sure we can hard code `Object` in this place exactly because it may produce false positives depending on the nesting. I think this needs to be a part of the constant resolution.

Take this scenario as an example (and let's add a unit test for it):

```ruby
module Foo
  class Object
  end

  # This `Bar` class inherits from `::Object`. But if we hard code it here to
  # `Object`, then it will be resolved to `Foo::Object`, which is a different
  # class
  class Bar
  end
end
```

Reply:
I think changing it to `::Object` actually works because we don't clear`::` for parent classes.

```rb
@index.index_single(IndexablePath.new(nil, "/fake/path/foo.rb"), <<~RUBY)
  module Foo
    class Object; end

    class Bar; end
    class Baz < Object; end
  end
RUBY

assert_equal(["Foo::Bar", "Object", "Kernel", "BasicObject"], @index.linearized_ancestors_of("Foo::Bar"))
assert_equal(
  ["Foo::Baz", "Foo::Object", "Object", "Kernel", "BasicObject"],
  @index.linearized_ancestors_of("Foo::Baz"),
)
```

Reply:
Good idea, but if we follow this approach, there are other things we should optimize too.

We can do it in a follow up PR, but it'd be nice to get it out shortly after.

1. Entry owners should no longer be nilable. Every method/attribute/instance variable must always have an owner because top level declarations are inserted into the `Object` class. You'll want to put `Object` as the initial entry in the owner stack and then get rid of the nilable aspect of owner everywhere
2. [This](https://github.com/Shopify/ruby-lsp/blob/bed467e9cdb842e3af6c7dc82d3ee2c9b99123ae/lib/ruby_indexer/lib/ruby_indexer/index.rb#L378) should be simplified, since every class will always have a parent class (careful with `BasicObject` since it needs to be treated as a special case). The nil checks below should also be removed